### PR TITLE
[Fixes #3774] DataAnnotation validation ignored

### LINKED ISSUE #3774: DataAnnotation validation ignored
DataAnnotation validation is ignored if route parameter has the same name as object variable passed in the body of the HTTP request

Model:

```c#
public class TransferInfo
{
        [Range(1, int.MaxValue)] //or CustomValidationAttribute
        public int AccountId { get; set; }
        public double Amount { get; set; }
}
```

Controller:

```c#
[HttpPost("{accountId:int}")]
public IActionResult Transfer(int accountId, [FromBody] TransferInfo transferInfo)
{
    if (!ModelState.IsValid)
    {
        return HttpBadRequest();
    }
    ...
}

```

Posted data (JSON):

```json
{
     "accountId": 0,
     "amount": 250.0
}
```

ModelState is valid even if I check that accountId must be greater than 0, using Range or Custom validation attribute.

If I update my TransferInfo class definition and rename "AccountId" by "ToAccountId" and try to send:

```json
{
     "toAccountId: 0,
     "amount": 250.0
}
```

ModelState Validation will failed

### PR DESCRIPTION
**Pull request by kichalla**
@rynowak

### CODE REVIEW

**Comment by kichalla** [comment]
@dougbu

File: src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs [hunk 1]
```diff
@@ -98,7 +98,7 @@ public bool Validate(ModelMetadata metadata, string key, object model)
         protected virtual bool ValidateNode()
         {
             var state = _modelState.GetValidationState(_key);
-            if (state == ModelValidationState.Unvalidated)
+            if (state != ModelValidationState.Invalid)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs]: Let's add a comment here. Something like

```
// Rationale: we might see the same model state key used for two different objects.
// We want to run validation unless it's already known that this key is invalid.
```

**Comment by rynowak** [comment]
:shipit: with a comment from me. Get signoff from @dougbu as well

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/ValidationIntegrationTests.cs [hunk 2]
```diff
@@ -18,6 +20,133 @@ namespace Microsoft.AspNetCore.Mvc.IntegrationTests
...
+        private class TestController { }
+
+        public static TheoryData<List<ParameterDescriptor>> MultipleActionParametersAndValidationData
+        {
+            get
+            {
+                return new TheoryData<List<ParameterDescriptor>>
+                {
+                    // Irrespective of the order in which the parameters are defined on the action,
+                    // the validation on the TransferInfo's AccountId should occur.
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/ValidationIntegrationTests.cs]: May be worth mentioning the `accountId` is bound using the prefix `"accountId"` while `transferInfo` is bound using the empty prefix (and the `TransferInfo` property names).

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/ValidationIntegrationTests.cs [hunk 3]
```diff
@@ -18,6 +20,133 @@ namespace Microsoft.AspNetCore.Mvc.IntegrationTests
...
+                actionDescriptor: actionDescriptor);
+
+            var modelState = operationContext.ActionContext.ModelState;
+
+            // Act
+            var arguments = await argumentBinder.BindActionArgumentsAsync(
+                (ControllerContext)operationContext.ActionContext, new TestController());
+
+            // Assert
+            Assert.True(modelState.IsValid);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/ValidationIntegrationTests.cs]: Confirm the values in the two arguments.

**Comment by dougbu** [comment]
:shipit: w/ slight test updates.

**Comment by kichalla** [comment]
7337f501120cfcd9fb46166b38c360a86282b879

**Pull request closed without merging**